File: src/main/java/org/eln2/sim/mna/component/Capacitor.java
package org.eln2.sim.mna.component;

/*
NOTE: DO NOT IMPORT MINECRAFT CODE IN THIS CLASS
EXTEND IT INSTEAD IN THE org.eln.nbt DIRECTORY
 */

import org.eln2.sim.mna.SubSystem;
import org.eln2.sim.mna.misc.ISubSystemProcessI;
import org.eln2.sim.mna.state.State;

public class Capacitor extends Bipole implements ISubSystemProcessI {

    private double c = 0;
    double cdt;

    public Capacitor() {
    }

    public Capacitor(State aPin, State bPin) {
        connectTo(aPin, bPin);
    }

    @Override
    public double getCurrent() {
        return 0;
    }

    public void setC(double c) {
        this.c = c;
        dirty();
    }

    @Override
    public void applyTo(SubSystem s) {
        cdt = c / s.getDt();

        s.addToA(getAPin(), getAPin(), cdt);
        s.addToA(getAPin(), getBPin(), -cdt);
        s.addToA(getBPin(), getBPin(), cdt);
        s.addToA(getBPin(), getAPin(), -cdt);
    }

    @Override
    public void simProcessI(SubSystem s) {
        double add = (s.getXSafe(getAPin()) - s.getXSafe(getBPin())) * cdt;
        s.addToI(getAPin(), add);
        s.addToI(getBPin(), -add);
    }

    @Override
    public void quitSubSystem() {
        getSubSystem().removeProcess(this);
        super.quitSubSystem();
    }

    @Override
    public void setSubSystem(SubSystem s) {
        super.setSubSystem(s);
        s.addProcess(this);
    }

    public double getE() {
        double u = getU();
        return u * u * c / 2;
    }

    public double getC() {
        return c;
    }
}

File: src/main/java/org/eln2/sim/mna/component/Capacitor.kt
package org.eln2.sim.mna.component

import org.eln2.sim.mna.SubSystem
import org.eln2.sim.mna.misc.ISubSystemProcessI
import org.eln2.sim.mna.state.State

/*
NOTE: DO NOT IMPORT MINECRAFT CODE IN THIS CLASS
EXTEND IT INSTEAD IN THE org.eln.nbt DIRECTORY
 */

open class Capacitor: Bipole, ISubSystemProcessI {

    constructor()
    constructor(aPin: State?, bPin: State?): super(aPin, bPin)

    open var c: Double = 0.0
        set(c) {
            field = c
            dirty()
        }

    override fun getI(): Double {
        // TODO: Uhhhhhhhhhhh
        // No joke, this is what the old code did.
        return 0.0
    }

    override fun applyTo(s: SubSystem) {
        val cdt = c / s.dt
        s.addToA(aPin, aPin, cdt)
        s.addToA(aPin, bPin, -cdt)
        s.addToA(bPin, bPin, cdt)
        s.addToA(bPin, aPin, -cdt)
    }

    override fun simProcessI(s: SubSystem?) {
        if (s== null) return
        val cdt = c / s.dt
        val add: Double = (s.getXSafe(aPin) - s.getXSafe(bPin)) * cdt
        s.addToI(aPin, add)
        s.addToI(bPin, -add)
    }

    override fun quitSubSystem() {
        subSystem!!.removeProcess(this)
        super.quitSubSystem()
    }

    open fun getE(): Double {
        val u = u
        return u * u * c / 2
    }
}